drivers: i3c: i3c/i2c attach/detach api

XenuIsWatching opened the pull request:
There are some needs to attach and reattach i3c/i2c devices at runtime
Some I2C devices can have special registers where the address can be
changed at runtime. Also some I3C devices can be powered off at runtime
freeing up the address space they take up.  These new APIs allow for these
to be changed at runtime. This also moves some config/data in to a common
i3c config/data structure which would allow the api to operate on to be
common for all I3C drivers.

Based/Depends on: https://github.com/zephyrproject-rtos/zephyr/pull/51538

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching commented:
Something I ought to consider is that the singly-linked lists functions are not thread-safe and will likely need spinlocks around it

XenuIsWatching commented:
I just realized that slist need a spinlock with them as they are not thread-safe

XenuIsWatching commented:
....and I also just realized that calls to the function `i3c_addr_slots_set` are also not thread-safe either due to the writes to `i3c_addr_slot` not being atomic

XenuIsWatching commented:
a note/warning was added to the api doxygen saying it is not thread-safe

dcpleung commented:
I think it might be better to extract the API and common area changes into its own PR. Then the CDN I3C driver can then be based on top of that. The API change is simple enough so that it can be merged on its own (even without corresponding driver implementation).

rderojas commented:
This is great! Thanks for adding this 🥇

XenuIsWatching commented:
> I think it might be better to extract the API and common area changes into its own PR. Then the CDN I3C driver can then be based on top of that. The API change is simple enough so that it can be merged on its own (even without corresponding driver implementation).

Sorry.... but git went haywire on me and you may have looked at the wrong commit :angry: 
Please take a look at https://github.com/zephyrproject-rtos/zephyr/commit/f17e795cb1420f1833561b45a2c0fb8cdb54cc7f

dcpleung commented:
What I meant was to extract the last two commits out (or even all commits other than the CDN driver) as its own PR. Then you don't have to introduce the CDN driver, and change again because of the de-/attach API.

XenuIsWatching committed:
drivers: i3c: cdns: add reattach api implementation

This adds the reattach api necessary for writing the i3c retaining
registers within the cdns i3c when the dynamic address changes.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
drivers: i3c: i3c attach/detach api

There are some needs to attach and reattach i3c/i2c devices at runtime
Some I2C devices can have special registers where the address can be
changed at runtime. Also some I3C devices can be powered off at runtime
freeing up the address space they take up. These new APIs allow for these
to be changed at runtime. This also moves some config/data in to a common
i3c config/data structure which would allow the api to operate on to be
common for all I3C drivers.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
drivers: i3c: cdns: add attach/detach api implementation

The cadence i3c ip requires it's retaining registers to be updated
when a device is detached or attached.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching commented:
force pushing up rebasing on top of main after https://github.com/zephyrproject-rtos/zephyr/pull/51538 has been merged in

dcpleung commented:
Hm... could you try rebasing and pushing it again? Seems like some CI actions were incomplete.

XenuIsWatching commented:
Found a few minor issues that have now been pushed up
* Error checking in `i3c_bus_setdasa` for `i3c_reattach_i3c_device` and `i3c_detach_i3c_device` was added
* Fix Doxygen Header Comments Copy/Paste Typos on `i3c_detach_i2c_device`
* Fix bug with `i3c_attach_i3c_device` address decision tree if `init_dynamic_addr` is set

dcpleung commented on include/zephyr/drivers/i3c.h:
```diff
@@ -1100,6 +1214,101 @@ static inline int i3c_recover_bus(const struct device *dev)
+ * @retval -EINVAL If address is not available or if the device
+ *     has already been attached before
+ */
+static inline int i3c_attach_i3c_device(struct i3c_device_desc *target)
```
Since this is marked `inline`, it will possibly be expanded at the caller site. Please add a helper function in C code for all this, and have this call the helper function instead. Ditto for the various additions below.

XenuIsWatching replied:
I assume you're referring to just put the "figuring out the DA" decision tree in a helper function. Yeah, I'll do that

dcpleung replied:
I am thinking the whole function body. This type of stubs are supposed to be really light weight and is to call the driver API directly (if API func exists).

XenuIsWatching replied:
ok...?? If I put the whole function within a function then there wouldn't be a point to the upper function still being around as you should just be calling the helper function which should always be implemented (and then it would call the driver api (if it exist)) :/ I understand your point about inline expanding a function that's a bit more than few lines, but doing that would just make the i3c api func unnecessary imo.

Would it make sense to remove these "inline" apis and just expect the application to call the helper functions?

dcpleung replied:
Oh yeah... just remove inline and leave it as a function prototype, and put implementation somewhere else.

(... my brain is not working today)

XenuIsWatching replied:
done

XenuIsWatching commented on include/zephyr/drivers/i3c.h:
```diff
@@ -1111,24 +1320,167 @@ static inline int i3c_recover_bus(const struct device *dev)
-		(const struct i3c_driver_api *)target->bus->api;
+	struct i3c_driver_data *data = (struct i3c_driver_data *)target->bus->data;
+	const struct i3c_driver_api *api = (const struct i3c_driver_api *)target->bus->api;
+	int status = 0;
```
That's not the intent. `api->reattach_i3c_device` is an optional api to adjust retaining registers (if the IP has it). Returning `-ENOSYS` could be confusing if using the Silvca (NXP) IP which does not require any of the retaining register updates. Essentially all this function would do is update the address slots if that address is free.

dcpleung replied:
The device API usually calls for `-ENOSYS` if the API is not implemented at the driver level. And there is (almost) always 1-to-1 relations between this type of stubs and corresponding APIs. Probably a good idea to mention this in the API doc if this still works without driver implementation.

XenuIsWatching replied:
I though `-ENOSYS` would be if the _"entire"_ function is not available which isn't exactly the case here. Id rather see 0 for the case of portability between switching between different IP if it is completely successful. I can add an extra note to the api doxygen about how it will call an optional api if implemented.

XenuIsWatching commented on include/zephyr/drivers/i3c.h:
```diff
@@ -1111,24 +1320,167 @@ static inline int i3c_recover_bus(const struct device *dev)
-		(const struct i3c_driver_api *)target->bus->api;
+	struct i3c_driver_data *data = (struct i3c_driver_data *)target->bus->data;
+	const struct i3c_driver_api *api = (const struct i3c_driver_api *)target->bus->api;
+	int status = 0;
```
duplicate comment?

dcpleung replied:
Probably because my internet was down momentarily after I pressed "add comment"... and it allowed me to press the button again. So I did and we have a dupe.

XenuIsWatching committed:
drivers: i3c: add reattach_i3c_device api

Some I3C controllers have retaining registers which are used to contain
the DA of the i3c device. This needs to be updated every time the DA is
updated with SETNEWDA or SETDASA

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
drivers: i3c: add SETNEWDA ccc helper function

This is a ccc helper function for setting a new dynamic addr for
a Target.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>
